instrumentation/jms/jms-1.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jms/v1_1/JmsMessageProducerInstrumentation.java

```diff
@@ -52,103 +52,90 @@ public void transform(TypeTransformer transformer) {
-      Destination defaultDestination;
-      try {
-        defaultDestination = producer.getDestination();
-      } catch (JMSException e) {
-        defaultDestination = null;
-      }
-
-      Context parentContext = Java8BytecodeBridge.currentContext();
-      request =
+      MessageWithDestination messageWithDestination =
           MessageWithDestination.create(
-              JavaxMessageAdapter.create(message),
-              JavaxDestinationAdapter.create(defaultDestination));
-      if (!producerInstrumenter().shouldStart(parentContext, request)) {
-        return;
+              JavaxMessageAdapter.create(message), JavaxDestinationAdapter.create(destination));
+      if (!producerInstrumenter().shouldStart(parentContext, messageWithDestination)) {
+        return new AdviceScope(callDepth, null, null, null);
       }
 
-      context = producerInstrumenter().start(parentContext, request);
-      scope = context.makeCurrent();
+      Context context = producerInstrumenter().start(parentContext, messageWithDestination);
+      return new AdviceScope(callDepth, messageWithDestination, context, context.makeCurrent());
     }
 
-    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
-    public static void stopSpan(
-        @Advice.Local("otelCallDepth") CallDepth callDepth,
-        @Advice.Local("otelRequest") MessageWithDestination request,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope,
-        @Advice.Thrown Throwable throwable) {
+    public void end(@Nullable Throwable throwable) {
       if (callDepth.decrementAndGet() > 0) {
         return;
       }
-
-      if (scope != null) {
-        scope.close();
-        producerInstrumenter().end(context, request, null, throwable);
+      if (scope == null || context == null || messageWithDestination == null) {
+        return;
       }
+
+      scope.close();
+      producerInstrumenter().end(context, messageWithDestination, null, throwable);
     }
   }
 
   @SuppressWarnings("unused")
-  public static class ProducerWithDestinationAdvice {
+  public static class ProducerAdvice {
 
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void onEnter(
-        @Advice.Argument(0) Destination destination,
-        @Advice.Argument(1) Message message,
-        @Advice.Local("otelCallDepth") CallDepth callDepth,
-        @Advice.Local("otelRequest") MessageWithDestination request,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
-      callDepth = CallDepth.forClass(MessageProducer.class);
-      if (callDepth.getAndIncrement() > 0) {
-        return;
+    public static AdviceScope onEnter(
+        @Advice.Argument(0) Message message, @Advice.This MessageProducer producer) {
+      Destination destination;
+      try {
+        destination = producer.getDestination();
+      } catch (JMSException e) {
+        destination = null;
```

Comment:
[for reviewer] for simplicity we execute this try/catch early in the advice. That could become an avoidable overhead if a thrown exception is likely for nested calls, in which case the nested call check should come first and they this try/catch, and that would imply creating a dedicated `start` method.